flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/calcite/Rank.scala

```diff
@@ -91,10 +89,13 @@ abstract class Rank(
   }
 
   override def deriveRowType(): RelDataType = {
+    if (!outputRankNumber) {
+      return input.getRowType
+    }
     val typeFactory = cluster.getRexBuilder.getTypeFactory
     val typeBuilder = typeFactory.builder()
     input.getRowType.getFieldList.foreach(typeBuilder.add)
-    // rank function column is always the last column, and its type is BIGINT NOT NULL
+    // rank number column is always the last column, and its type is BIGINT NOT NULL
     val allFieldNames = new util.HashSet[String]()
     allFieldNames.addAll(input.getRowType.getFieldNames)
     val rankFieldName = FlinkRelOptUtil.buildUniqueFieldName(allFieldNames, "rk")
```

Comment:
Why we can force the rank field name to "rk"? What if the user specified the name for it? Like:
SELECT * FROM (
  SELECT a, b, RANK() OVER (PARTITION BY b ORDER BY c) `myRank` FROM MyTable
) t
WHERE `myRank` < 10 ?

Reply:
yes, the rank number field name should be the user specified name

Reply:
actually, I find that the rank number field name may be dropped in `ProjectToWindowRule`, and be changed to name like `w0$oi`. so, I think we can keep current implementation and add a `TODO` to solve this later.

Reply:
Not sure about this, it looks like a BLOCKER to me. Without this, i don't think user can simply write a correct query to meet this optimization.